openpype/hosts/cinema4d/plugins/publish/collect_instances.py

```diff
@@ -0,0 +1,185 @@
+            members = [lib.ObjectPath(obj=members.ObjectFromIndex(doc, idx)) for idx in range(members.GetObjectCount())]
+
+
+            # Collect Children - c4d.BaseObject
+            children = [lib.ObjectPath(obj=obj) for member in members for obj in lib.walk_hierarchy(member.obj.GetDown())]
+
+            parents = []
+            if data.get("includeParentHierarchy", True):
+                # If `includeParentHierarchy` then include the parents
+                # so they will also be picked up in the instance by validators
+                parents = [lib.ObjectPath(obj=obj) for obj in self.get_all_parents(members)]
+            members_hierarchy = list(
+                    set(
+                        [str(x) for x in members] + \
+                        [str(x) for x in children] + \
+                        [str(x) for x in parents]
+                    )
+                )
+
+            if 'families' not in data:
+                data['families'] = [data.get('family')]
+
+            # Create the instance
+            instance = context.create_instance(objset.op.GetName())
+            instance[:] = members_hierarchy
+
+            # Store the exact members of the object set
+            instance.data["setMembers"] = [lib.serialize_c4d_data(x) for x in members]
+
+            # Define nice label
+            name = objset.op.GetName()  # use short name
+            label = "{0} ({1})".format(name,
+                                       data["asset"])
+
+            # Append start frame and end frame to label if present
+            if "frameStart" and "frameEnd" in data:
+
+                # if frame range on maya set is the same as full shot range
+                # adjust the values to match the asset data
+                if (ctx_frame_start_handle == data["frameStart"]
+                        and ctx_frame_end_handle == data["frameEnd"]):  # noqa: W503, E501
+                    data["frameStartHandle"] = ctx_frame_start_handle
+                    data["frameEndHandle"] = ctx_frame_end_handle
+                    data["frameStart"] = ctx_frame_start
+                    data["frameEnd"] = ctx_frame_end
+                    data["handleStart"] = ctx_handle_start
+                    data["handleEnd"] = ctx_handle_end
+
+                # if there are user values on start and end frame not matching
+                # the asset, use them
+
+                else:
+                    if "handles" in data:
+                        data["handleStart"] = data["handles"]
+                        data["handleEnd"] = data["handles"]
+                    else:
+                        data["handleStart"] = 0
+                        data["handleEnd"] = 0
+
+                    data["frameStartHandle"] = data["frameStart"] - data["handleStart"]  # noqa: E501
+                    data["frameEndHandle"] = data["frameEnd"] + data["handleEnd"]  # noqa: E501
+
+                if "handles" in data:
+                    data.pop('handles')
```

Comment:
FYI - This logic can probably be greatly simplified, see #3437 - even though that one is still in discussion.